tests/test_browser.py

```diff
@@ -294,5 +299,49 @@ def test_post(httpbin):
     assert resp.status_code == 200 and resp.json()['form'] == data
 
 
+def test_encoding(httpbin):
+    url = 'mock://encoding'
+    text = (
+        '<!doctype html>'
+        '<html lang="fr">'
+        '<head><meta charset="utf-8"><title>Titleéàè</title></head>'
+        '<body></body>'
+        '</html>'
+    )
+
+    # Check page with UTF-8 charset in HTML and no encoding in HTTP headers
+    browser, adapter = prepare_mock_browser()
+    mock_get(adapter, url=url, reply=text)
+    browser.open(url)
+    assert browser.page.original_encoding == 'utf-8'
+
+    # Check page with UTF-8 charset in HTML and utf-8 in HTTP headers
+    mock_get(
+        adapter,
+        url=url,
+        reply=text,
+        content_type='text/html; charset=utf-8'
+    )
+    browser.open(url)
+    assert browser.page.original_encoding == 'utf-8'
+
+    # Check page with no charset in HTML and utf-8 in HTTP headers
+    text = (
+        '<!doctype html>'
+        '<html lang="fr">'
+        '<head><title>Titleéàè</title></head>'
+        '<body></body>'
+        '</html>'
+    )
+    mock_get(
+        adapter,
+        url=url,
+        reply=text,
+        content_type='text/html; charset=utf-8'
+    )
```

Comment:
I think this lacks the "both HTML and HTTP specify an encoding, but they are in conflict", which IMHO is an important one. Can you add it? (that's an argument in favor of parameterization, if you refactor before, adding a case is just a matter of adding one item to the parameters list)

Reply:
I can definitely add this test, but I'm not sure what should be the correct behavior here. I would consider such conflict to be a bug on the remote side.

Not sure which one should have precedence. It is currently `html_encoding` but maybe it should be `http_encoding`?

https://github.com/MechanicalSoup/MechanicalSoup/pull/355/files#diff-8ddd0e386c398170601ea5f855d66494d04be5752fc7c88b12af0fd3bb9f48e8R75

Reply:
In any case, testing this is a way to document the current behavior and make sure we don't change it without noticing.

What do web browsers do in this case? I think they favor the HTTP header too (and IIRC I've already be bitten by this with my sysadmin setting the HTTP header I had no control on, and me trying to set the encoding in an HTML static page), but I didn't check.